Building CLI Applications (#372)
crystal-lang/crystal-book: overview/cli.md

# Command Line Interface Application

Programming Command Line Interface applications (CLI applications) is one of the most entertaining tasks a developer may do. So let’s have some fun building our first CLI application in Crystal.

There are two main topics when building a CLI application:
* [input](#input)
* [output](#output)

## Input
This topic covers all things related to:
* [options passed to the app](#options)
* [request for user input](#request-for-user-input)

### Options
It is a very common practice to pass options to the application. For example, we may run `crystal -v` and Crystal will display:

```shell-session
$ crystal -v
Crystal 0.31.1 (2019-10-02)

LLVM: 8.0.1
Default target: x86_64-apple-macosx
```

and if we run: `crystal -h`, then Crystal will show all the accepted options and how to use them.

So now the question would be: **do we need to implement an options parser?** No need to, Crystal got us covered with the class `OptionParser`. Let’s build an application using this parser!

> To begin with our CLI application will have two options:

Reviewer: ```suggestion
At start our CLI application will have two options:
```

We can read all about `OptionParser` in [the official API documentation](https://crystal-lang.org/api/latest/OptionParser.html). And from there we are one click away from the source code ... the actual proof that it is not magic!

Now, let's run our application. We have two ways [using the compiler](https://crystal-lang.org/reference/using_the_compiler):
1. [Build the application](https://crystal-lang.org/reference/using_the_compiler/#crystal-build) and then run it.
2. Compile and [run the application](https://crystal-lang.org/reference/using_the_compiler/#crystal-run), all in one command.

We are going to use the second way:

```shell-session
$ crystal ./all_my_cli.cr -- -h

Welcome to The Beatles App!
    -v, --version                    Show version
    -h, --help                       Show help
```

Let's continue building our _fabulous_ application adding a new feature:

> by default (i.e. no options given) the application will display the names of the Fab Four. But, if we pass the option `-t` / `--twist` it will display the names in uppercase:

Reviewer: ```suggestion
By default (i.e. no options given) the application will display the names of the Fab Four. But, if we pass the option `-t` / `--twist` it will display the names in uppercase:
```

```crystal
# file: all_my_cli.cr
require "option_parser"

the_beatles = [
  "John Lennon",
  "Paul McCartney",
  "George Harrison",
  "Ringo Starr"
]
shout = false

option_parser = OptionParser.parse do |parser|
  parser.banner = "Welcome to The Beatles App!"

  parser.on "-v", "--version", "Please, -please- show me the version" do
    puts "version 1.0"
    exit
  end
  parser.on "-h", "--help", "Show help" do
    puts parser
    exit
  end
  parser.on "-t", "--twist", "Twist and SHOUT" do
    shout = true
  end
end

members = the_beatles
members = the_beatles.map &.upcase if shout

puts ""
puts "Group members:"
puts "=============="
members.each do |member|
  puts member
end
```

Running the application with the `-t` option will output:

```shell-session
$ crystal run ./all_my_cli.cr -- -t

Group members:
==============
JOHN LENNON
PAUL MCCARTNEY
GEORGE HARRISON
RINGO STARR
```

#### Parameterized options

> Let’s add a new feature to our application: _when passing the option -g, the application will say hello to a given name **passed as a parameter to the option**_.

Reviewer: ```suggestion
Let’s add a new feature to our application: _when passing the option `-g` / `--goodbye_hello`, the application will say hello to a given name **passed as a parameter to the option**_.
```

```crystal
option_parser = OptionParser.parse do |parser|
  #...

  parser.on "-g NAME", "--goodbye_hello=NAME", "Say hello to whoever you want" do |name|
    say_hi_to = name
  end

  #... etc
```

In this case, the block receives a parameter that represents the parameter passed to the option.

> **Note:** Before running the application, for this new option to be functional, some other changes need to be made. See [the final result](#the-complete-application).

Reviewer: Maybe the above code example could be a minimal but working application demonstrating the `-g` option.
```cr
say_hi_to = nil
option_parser = OptionParser.parse do |parser|
	  parser.on "-g NAME", "--goodbye_hello=NAME", "Say hello to whoever you want" do |name|
    say_hi_to = name
  end
end
unless say_hi_to.empty?
  puts ""
  puts "You say goodbye, and I say hello to #{say_hi_to}!"
end
```

  end
  parser.missing_option do |option_flag|
    STDERR.puts "ERROR: #{option_flag} is missing something."
    STDERR.puts ""
    STDERR.puts parser
    exit(1)
  end
  parser.invalid_option do |option_flag|
    STDERR.puts "ERROR: #{option_flag} is not a valid option."
    STDERR.puts parser
    exit(1)
  end
end

members = the_beatles
members = the_beatles.map &.upcase if shout

puts "Strawberry fields forever mode ON" if strawberry

puts ""
puts "Group members:"
puts "=============="
members.each do |member|
  puts "#{strawberry ? "🍓" : "-"} #{member}"
end

unless say_hi_to.empty?
  puts ""
  puts "You say goodbye, and I say hello to #{say_hi_to}!"
end
```

### Request for user input

Sometimes, we may need the user to input a value. How do we _read_ that value?
Easy, peasy! Let’s create a new application: the Fab Four will sing with us any phrase we want. When running the application, it will request a phrase to the user and the magic will happen!

```crystal
# file: let_it_cli.cr
puts "Welcome to The Beatles Sing Along version 1.0!"
puts "Enter a phrase you want The Beatles to sing"
print "> "
user_input = gets
puts "The Beatles are singing: 🎵#{user_input}🎶🎸🥁"
```

> The method `gets` will **pause** the execution of the application, until the user finishes entering the input (pressing the `Enter` key)

Reviewer: ```suggestion
The method [`gets`](https://crystal-lang.org/api/latest/toplevel.html#gets(*args,**options)-class-method) will **pause** the execution of the application, until the user finishes entering the input (pressing the `Enter` key).
```

When the user presses `Enter`, then the execution will continue and `user_input` will have the user value.

> But what happen if the user doesn’t enter any value? In that case, we would  get an `empty string` (if the user only presses `Enter`) or maybe a `Nil value`(if the user hits Ctrl+D).

Reviewer: ```suggestion
But what happen if the user doesn’t enter any value? In that case, we would  get an empty string (if the user only presses `Enter`) or maybe a `Nil` value (if the input stream id closed, e.g. by pressing `Ctrl+D`).
```

So, we have to test for `Nil` and for `empty` and act naturally for each case:

```crystal
# file: let_it_cli.cr
puts "Welcome to The Beatles Sing Along version 1.0!"
puts "Enter a phrase you want The Beatles to sing"
print "> "
user_input = gets

exit if user_input.nil? # Ctrl+D

default_lyrics = "Na, na, na, na-na-na na" \
                 " / " \
                 "Na-na-na na, hey Jude"

lyrics = user_input.presence || default_lyrics

puts "The Beatles are singing: 🎵#{lyrics.upcase}🎶🎸🥁"
```

## Output

Now, we will focus on the second main topic: our application’s output.
For starters, our applications already display information but (I think) we could do better. Let’s add more _life_ (i.e. colors!) to the outputs.

And to accomplish this, we will be using the [`Colorize`](https://crystal-lang.org/api/latest/Colorize.html) module.

Let’s focus on our first application. We will add colors to the application’s banner. We will use white font on a black background:

```crystal
# file: all_my_cli.cr
require "option_parser"
require "colorize"

# ...

option_parser = OptionParser.parse do |parser|

>   parser.banner = "#{"The Beatles".colorize(:white).on(:black)} App"

Reviewer: Maybe make this example really tiny and self-contained?
 ```suggestion
puts "#{"The Beatles".colorize(:white).on(:black)} App"
```

    shout = true
  end
end

members = the_beatles
members = the_beatles.map &.upcase if shout

puts ""
puts "Group members:"
puts "=============="
members.each do |member|
  puts member
end
```

Running the application with the `-t` option will output:

```shell-session
$ crystal run ./twist_and_shout.cr -- -t

Group members:
==============
JOHN LENNON
PAUL MCCARTNEY
GEORGE HARRISON
RINGO STARR
```

#### Parameterized options
Let’s create another application: _when passing the option `-g` / `--goodbye_hello`, the application will say hello to a given name **passed as a parameter to the option**_.

```crystal
# file: hello_goodbye.cr
require "option_parser"

the_beatles = [
  "John Lennon",
  "Paul McCartney",
  "George Harrison",
  "Ringo Starr"
]
say_hi_to = ""

option_parser = OptionParser.parse do |parser|
  parser.banner = "Welcome to The Beatles App!"

  parser.on "-v", "--version", "Show version" do
    puts "version 1.0"
    exit
  end
  parser.on "-h", "--help", "Show help" do
    puts parser
    exit
  end
  parser.on "-g NAME", "--goodbye_hello=NAME", "Say hello to whoever you want" do |name|
    say_hi_to = name
  end
end

unless say_hi_to.empty?
  puts ""

>   puts "You say goodbye, and #{the_beatles[Random.new.rand(4)]} say hello to #{say_hi_to}!"

Reviewer: ```suggestion
  puts "You say goodbye, and #{the_beatles[Random.new.rand(4)]} says hello to #{say_hi_to}!"
```

  parser.on "-v", "--version", "Please, -please- show me the version" do
    puts "version 1.0"
    exit
  end
  parser.on "-h", "--help", "Help! I need somebody..." do
    puts parser
    exit
  end
  parser.on "-t", "--twist", "Twist and SHOUT" do
    shout = true
  end
end

members = the_beatles
members = the_beatles.map &.upcase if shout

puts ""
puts "Group members:"
puts "=============="
members.each do |member|
  puts member
end
```

Furthermore, we may want to **pass a parameter to an option**. Let’s add a new option to our application: when passing this option, the application will say hello to the given name (passed as a parameter to the option). Here we go:

First we define a new variable:

```crystal
say_hi_to = ""

option_parser = OptionParser.parse do |parser|
  # ... etc
```

Then, we define the new option:

```crystal
option_parser = OptionParser.parse do |parser|
  parser.banner = "Welcome to The Beatles App!"

  parser.on "-g NAME", "--goodbye_hello=NAME", "Say hello to whoever you want" do |name|
    say_hi_to = name
  end

  # ... etc
```

Finally, after listing the group members, we add:

```crystal
members.each do |member|
  puts member
end

> if (!say_hi_to.empty?)

Reviewer: More idiomatic: 

```suggestion
unless !say_hi_to.empty?
```

 ```crystal
# file: let_it_cli.cr
puts "Welcome to The Beatles Sing Along version 1.0!"
puts "Enter a phrase you want The Beatles to sing"
print "> "
user_input = gets
puts "The Beatles are singing: 🎵#{user_input.upcase}🎶🎸🥁"
```

When running the example, Crystal will reply:

```bash
$ crystal ./let_it_cli.cr
Showing last frame. Use --error-trace for full trace.

In let_it_cli.cr:5:46

 5 | puts "The Beatles are singing: 🎵#{user_input.upper_case}
                                                  ^---------
Error: undefined method 'upper_case' for Nil (compile-time type is (String | Nil))
```

Ah! We should have known better: the type of the user input is the [union type](https://crystal-lang.org/reference/syntax_and_semantics/type_grammar.html) `String | Nil`.
So, we have to test for `Nil`, and for `empty` and act naturally for each case:

```crystal
# file: let_it_cli.cr
puts "Welcome to The Beatles Sing Along version 1.0!"
puts "Enter a phrase you want The Beatles to sing"
print "> "
user_input = gets

exit if user_input.nil? # Ctrl+D

default_lyrics = "Na, na, na, na-na-na na" \
                 " / " \
                 "Na-na-na na, hey Jude"

> lyrics = user_input.presence? || default_lyrics

Reviewer: The method name is `String#presence`

```suggestion
lyrics = user_input.presence || default_lyrics
```

puts "The Beatles are singing: 🎵#{lyrics.upcase}🎶🎸🥁"
```

> ## Output

Reviewer: Maybe call this section "Colored Output"?

Now, we will focus on the second main topic: our application’s output.
For starters, our applications already display information but (I think) we could do better. Let’s add more _life_ (i.e. colors!) to the outputs.

> And to accomplish this, we will be using the `Colorize` module.

Reviewer: ```suggestion
And to accomplish this, we will be using the [`Colorize`](https://crystal-lang.org/api/latest/Colorize.html) module.
```

Let’s focus on our first application. We will add colors to the application’s banner. We will use white font on a black background:

```crystal

> app_banner = "#{"The Beatles".colorize(:white).on(:black)} App"

Reviewer: This example misses `require "colorize"`.

```

Great! That was easy!

And for our second application, we will add a text decoration (`blink`in this case):

```crystal
# file: let_it_cli.cr
require "colorize"

puts "Welcome to The Beatles Sing Along version 1.0!"
puts "Enter a phrase you want The Beatles to sing"
print "> "
user_input = gets

exit if user_input.nil? # Ctrl+D

default_lyrics = "Na, na, na, na-na-na na" \
                 " / " \
                 "Na-na-na na, hey Jude"

lyrics = user_input.presence? || default_lyrics

puts "The Beatles are singing: #{"🎵#{user_input}🎶🎸🥁".colorize.mode(:blink)}"
```

Let’s try the renewed application … and _hear_ the difference!!

And **now** we have two fabulous apps!!

You may read more about the `Colorize` module  [in the official API documentation](https://crystal-lang.org/api/latest/Colorize.html) and find a list of available colors and text decorations.

## Testing

As with any other application, at some point we would like to [write tests](https://crystal-lang.org/reference/guides/testing.html) for the different features.

> Right now the code containing the logic of each of the applications always gets executed with the `OptionParser`, i.e. there is no way to include that file without running the whole application. So first we would need to refactor the code, separating the code necessary for parsing options from the logic. Ones the refactor is done, then we could start testing the logic, and so including the file with the logic in the testing files we need. We leave this as an exercise for the reader.

Reviewer: ```suggestion
Right now the code containing the logic of each of the applications always gets executed with the `OptionParser`, i.e. there is no way to include that file without running the whole application. So first we would need to refactor the code, separating the code necessary for parsing options from the logic. Once the refactor is done, we could start testing the logic and including the file with the logic in the testing files we need. We leave this as an exercise for the reader.
```

## Using `Readline` and `NCurses`

In case we want to build richer CLI applications, there are libraries that can help us. Here we will name two well-known libraries: `Readline` and` NCurses`.

As stated in the documentation for the [GNU Readline Library](http://www.gnu.org/software/readline/), `Readline` is a library that provides a set of functions for use by applications that allow users to edit command lines as they are typed in.
`Readline` has some great features: filename autocompletion out of the box; custom autocompletion method; keybinding, just to mention a few. If we want to try it then the [crystal-lang/crystal-readline](https://github.com/crystal-lang/crystal-readline) shard will give us an easy API to use `Readline`.

On the other hand, we have `NCurses`(New Curses). This library allows developers to create _graphical_ user interfaces in the terminal. As its name implies, it is an improved version of the library named `Curses`, which was developed to support a text-based dungeon-crawling adventure game called Rogue!

> As you can imagine, there are already a couple of shards in the ecosystem that will allow us to use `NCurses` in Crystal!

Reviewer: Would be nice to link to those shards if they're already mentioned.

Author: yep, I totally agree! but i didn't test any of those libraries ... yet :), so i didn't want to mention any in particular.

Author: Btw, thanks for all the other suggestions!

Reviewer: Thanks for the writing! =)

I suppose it would be fine to link available shards and mention that they're not tested. But at least the interested reader gets a direct reference where to look. IMO that's better than just vaguely mentioning "there are some shards (somewhere in the interweb)".

Reviewer: Or at least a link to https://crystalshards.org/?filter=ncurses where you can discover these shards.
